.github/workflows/ci.yml

```diff
@@ -114,6 +114,12 @@ jobs:
           - "3.10"
           - "3.11"
           - "3.12"
+        exclude:
+          # MacOS Arm supports Python 3.10+
+          - os: MacOS
+            python: "3.8"
+          - os: MacOS
+            python: "3.9"
```

Comment:
Is there no way to test 3.8 and 3.9 against Mac?

Reply:
python-versions does not support 3.8 and 3.9 (https://github.com/actions/setup-python/issues/808) for macos-14 runner, seems there is a PR to help fix it currently open: https://github.com/actions/python-versions/pull/259

Reply:
Seems this partly caused by old versions of Python not bundling new versions of pip, which was recently discussed: https://discuss.python.org/t/update-strategy-for-pip-in-ensurepip-for-security-maintenance-releases

Reply:
You can re-`include` 3.8 and 3.9 on `macos-13` or `macos-12`, for example:

https://github.com/python/peps/pull/3763

Or like this:

https://github.com/hugovk/pepotron/pull/63

This allows us to keep testing 3.8+, covering both architectures, and unlike https://github.com/pypa/pip/pull/12593, doesn't ~double the number of macOS jobs.

Reply:
@hugovk that's what my original PR does: https://github.com/pypa/pip/pull/12593

I created this PR as an alternative in case pip maintainers want to run `macos-latest` only, rather than adding multiple macos OSes. Particularly because pip maintainers have expressed frustration with how long CI runs take.

Reply:
Old versions of virtualenv are seemingly not compatible with macos-14, I removed the restriction on virtualenv which seems to have a long sordid history last updated in https://github.com/pypa/pip/pull/11288

Most integration tests now pass but on fails (seemingly on macos-14 only):

```
2024-04-25T16:21:21.1996150Z [31mFAILED[0m tests/functional/test_uninstall_user.py::[1mTests_UninstallUserSite::test_uninstall_editable_from_usersite[0m - AssertionError: Script result: python -m pip install --user -e /private/var/folders/3m/p59k4qdj0f17st0gn2cmj3640000gn/T/pytest-of-runner/pytest-1/popen-gw2/test_uninstall_editable_from_u0/data/packages/FSPkg
2024-04-25T16:21:21.1996700Z -- stdout: --------------------
2024-04-25T16:21:21.1998990Z Obtaining file:///private/var/folders/3m/p59k4qdj0f17st0gn2cmj3640000gn/T/pytest-of-runner/pytest-1/popen-gw2/test_uninstall_editable_from_u0/data/packages/FSPkg
2024-04-25T16:21:21.1999880Z   Preparing metadata (setup.py): started
2024-04-25T16:21:21.2000610Z   Preparing metadata (setup.py): finished with status 'done'
2024-04-25T16:21:21.2000920Z Installing collected packages: FSPkg
2024-04-25T16:21:21.2001380Z   Running setup.py develop for FSPkg
2024-04-25T16:21:21.2001540Z Successfully installed FSPkg
2024-04-25T16:21:21.2001540Z 
2024-04-25T16:21:21.2001680Z -- created: -------------------
2024-04-25T16:21:21.2001770Z   venv/user/lib/python3.9
2024-04-25T16:21:21.2001860Z                   site-packages
2024-04-25T16:21:21.2002000Z                       FSPkg.egg-link  (146 bytes)
2024-04-25T16:21:21.2002100Z -- updated: -------------------
2024-04-25T16:21:21.2002160Z   venv/user/lib
```

I have no way to reproduce this locally, and I am unsure what the issue is.

Reply:
@uranusjr Unless this happens to be a flaky test one of the following needs to happen:

* Someone else needs to fix the test (I have no idea where to start and can't reproduce)
* Python 3.8 and 3.9 need excluding
* An alternative PR like https://github.com/pypa/pip/pull/12593 needs accepting which runs 3.8 and 3.9 for x86 but not for ARM